src/txmempool.cpp

```diff
@@ -1140,3 +1153,49 @@ std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept
     }
     assert(false);
 }
+
+std::vector<CTxMemPool::txiter> CTxMemPool::CalculateCluster(const std::vector<uint256>& txids) const
+{
+    AssertLockHeld(cs);
+    std::vector<txiter> cluster{GetIterVec(txids)};
+    if (cluster.size() != txids.size()) {
+        // We can't continue because the caller specified a tx that doesn't exist in the mempool.
+        // Return an empty vector to let them know this failed.
+        return {};
+    }
+    // Reserve total ancestor + descendant counts of each transaction.  This is an approximation; it
+    // may overestimate because transactions may share ancestors/descendants, and may underestimate
+    // because the cluster may include more than just ancestors and descendants.
+    cluster.reserve(std::accumulate(cluster.cbegin(), cluster.cend(), 0, [](size_t sum, const auto it) {
+        return sum + it->GetCountWithAncestors() + it->GetCountWithDescendants() - 1; }));
+    {
+        // Use epoch: visiting an entry means we have added it to the cluster vector. It does not
+        // necessarily mean the entry has been processed.
+        WITH_FRESH_EPOCH(m_epoch);
+        for (const auto& it : cluster) {
+            visited(it);
+        }
+        // i = index of where the list of entries to process starts
+        for (size_t i{0}, to_process_count{txids.size()}; i < to_process_count; ++i) {
+            // DoS protection: if not finished after processing 500 entries, just quit.
+            if (to_process_count > 500) return {};
```

Comment:
Maybe this 500 can be a `constexpr` in the `MiniMiner` class? I probably wouldn't suggest this if this is the only place it occurs, but the tests specify this value too. It might be nice to be able to change this limit by changing a single line of code.

Reply:
Disagree with txmempool getting this value from mini_miner, as that would create a circular dependency.
This kind of constant should live in src/kernel/mempool_options.h.